File: tray-app/MailMCP/ViewModels/StatusViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
public final class StatusViewModel {
    public private(set) var status: DaemonStatus?
    public private(set) var pendingApprovalCount: Int = 0
    public private(set) var lastError: String?

    private let client: IpcClient
    private var pollTask: Task<Void, Never>?
    private var notifTask: Task<Void, Never>?

    public init(client: IpcClient) {
        self.client = client
    }

    public func start() {
        pollTask = Task { [weak self] in await self?.pollLoop() }
        notifTask = Task { [weak self] in await self?.notificationLoop() }
    }

    public func stop() {
        pollTask?.cancel()
        notifTask?.cancel()
    }

    public func refresh() async {
        do {
            status = try await client.call("status")
            let approvals: [PendingApproval] = try await client.call("approvals.list")
            pendingApprovalCount = approvals.count
            lastError = nil
        } catch {
            lastError = error.localizedDescription
        }
    }

    private func pollLoop() async {
        while !Task.isCancelled {
            await refresh()
            try? await Task.sleep(for: .seconds(30))
        }
    }

    private func notificationLoop() async {
        do {
            let stream = try await client.subscribe(events: [
                "approval.requested",
                "approval.resolved",
                "account.added",
                "account.removed",
                "mcp.paused_changed",
            ])
            for await note in stream {
                switch note {
                case .approvalRequested:
                    pendingApprovalCount += 1
                case .approvalResolved:
                    pendingApprovalCount = max(0, pendingApprovalCount - 1)
                case .accountAdded, .accountRemoved, .mcpPausedChanged:
                    await refresh()
                default:
                    break
                }
            }
        } catch {
            lastError = error.localizedDescription
        }
    }
}

File: tray-app-win/MailMCP/ViewModels/StatusViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using MailMCP.IPC;

namespace MailMCP.ViewModels;

/// <summary>
/// Polls the daemon for status every 30 s and reacts to notifications. Properties
/// are observable so the tray menu / status bar text rebinds automatically when
/// state changes.
/// </summary>
public partial class StatusViewModel : ObservableObject
{
    private readonly IpcClient _client;
    private CancellationTokenSource? _cts;
    private Task? _pollTask;
    private Task? _notifTask;

    [ObservableProperty] private DaemonStatus? _status;
    [ObservableProperty] private int _pendingApprovalCount;
    [ObservableProperty] private string? _lastError;

    public StatusViewModel(IpcClient client) { _client = client; }

    public void Start()
    {
        _cts = new CancellationTokenSource();
        _pollTask = Task.Run(() => PollLoop(_cts.Token));
        _notifTask = Task.Run(() => NotificationLoop(_cts.Token));
    }

    public async Task StopAsync()
    {
        _cts?.Cancel();
        try
        {
            if (_pollTask is not null) await _pollTask.ConfigureAwait(false);
            if (_notifTask is not null) await _notifTask.ConfigureAwait(false);
        }
        catch (OperationCanceledException) { }
    }

    public async Task RefreshAsync(CancellationToken ct = default)
    {
        try
        {
            Status = await _client.CallAsync<DaemonStatus>("status", ct: ct).ConfigureAwait(false);
            var pending = await _client.CallAsync<PendingApproval[]>("approvals.list", ct: ct).ConfigureAwait(false);
            PendingApprovalCount = pending.Length;
            LastError = null;
        }
        catch (Exception ex)
        {
            LastError = ex.Message;
        }
    }

    private async Task PollLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            await RefreshAsync(ct).ConfigureAwait(false);
            try { await Task.Delay(TimeSpan.FromSeconds(30), ct).ConfigureAwait(false); }
            catch (OperationCanceledException) { return; }
        }
    }

    private async Task NotificationLoop(CancellationToken ct)
    {
        try
        {
            await foreach (var note in _client.SubscribeAsync(
                new[]
                {
                    "approval.requested",
                    "approval.resolved",
                    "account.added",
                    "account.removed",
                    "mcp.paused_changed",
                }, ct).ConfigureAwait(false))
            {
                switch (note)
                {
                    case DaemonNotification.ApprovalRequested:
                        PendingApprovalCount += 1;
                        break;
                    case DaemonNotification.ApprovalResolved:
                        PendingApprovalCount = Math.Max(0, PendingApprovalCount - 1);
                        break;
                    case DaemonNotification.AccountAdded:
                    case DaemonNotification.AccountRemoved:
                    case DaemonNotification.McpPausedChanged:
                        await RefreshAsync(ct).ConfigureAwait(false);
                        break;
                }
            }
        }
        catch (OperationCanceledException) { /* graceful */ }
        catch (Exception ex) { LastError = ex.Message; }
    }
}
